packages/cloud_firestore/cloud_firestore_platform_interface/lib/cloud_firestore_platform_interface.dart

```diff
@@ -0,0 +1,219 @@
+      }
+    }
+    _instance = instance;
+  }
+
+  /// This method ensures that [CloudFirestorePlatform] isn't implemented with `implements`.
+  ///
+  /// See class docs for more details on why using `implements` to implement
+  /// [CloudFirestorePlatform] is forbidden.
+  ///
+  /// This private method is called by the [instance] setter, which should fail
+  /// if the provided instance is a class implemented with `implements`.
+  void _verifyProvidesDefaultImplementations() {}
+
+  // Actual API 
+  // Global
+  /// Removes any listener by its handle.
+  /// All handles must be unique across al types of listeners.
+  Future<void> removeListener(int handle) async {
+    throw UnimplementedError('removeListener() is not implemented');
+  }
+
+  // Firestore
+  Future<void> enablePersistence(String app, {@required bool enable}) async {
+    throw UnimplementedError('enablePersistence() is not implemented');
+  }
+
+  Future<void> settings(String app, {
+      bool persistenceEnabled,
+      String host,
+      bool sslEnabled,
+      int cacheSizeBytes,
+    }) async {
+    throw UnimplementedError('settings() is not implemented');
+  }
+
+  Future<Map<String, dynamic>> runTransaction(String app, {
+    @required PlatformTransactionHandler transactionHandler,
+    int transactionTimeout,
+  }) async {
+    throw UnimplementedError('runTransaction() is not implemented');
+  }
+
+  // Document Reference
+  Future<void> setDocumentReferenceData(String app, {
+    @required String path,
+    Map<String, dynamic> data,
+    // TODO: Type https://firebase.google.com/docs/reference/js/firebase.firestore.SetOptions.html 
+    Map<String, dynamic> options,
+  }) async {
+    throw UnimplementedError('setDocumentReferenceData() is not implemented');
+  }
+
+  Future<void> updateDocumentReferenceData(String app, {
+    @required String path,
+    Map<String, dynamic> data,
+  }) async {
+    throw UnimplementedError('updateDocumentReferenceData() is not implemented');
+  }
+
+  // TODO: Type this return
+  Future<Map<String, dynamic>> getDocumentReference(String app, {
+    @required String path,
+    @required String source,
+  }) async {
+    throw UnimplementedError('getDocumentReference() is not implemented');
+  }
+
+  Future<void> deleteDocumentReference(String app, {
+    @required String path,
+  }) async {
+    throw UnimplementedError('deleteDocumentReference() is not implemented');
+  }
+
+  // TODO: Port to stream
+  Future<int> addDocumentReferenceSnapshotListener(String app, {
+    @required String path,
+    bool includeMetadataChanges,
+  }) async {
+    throw UnimplementedError('addDocumentReferenceSnapshotListener() is not implemented');
+  }
+
+  // Query
+  Stream<dynamic> snapshots(String app, {
+    @required String path,
+    bool isCollectionGroup,
+    Map<String, dynamic> parameters,
+    bool includeMetadataChanges,
+  }) {
+    throw UnimplementedError('snapshots() is not implemented');
+  }
+
+  //TODO: Type this return
```

Comment:
```suggestion
  // TODO(ditman): Type this return
```

Reply:
I need to look at the API docs, and I'll add all these types before the CR is GM :P